Select legal operand for "m" and "ZC" constraints

milica-lazarevic opened the pull request:
For "m" and "ZC" constraints, 9-bit offsets can be handled. Also, the two least-significant bits should be cleared as required by ll and sc instructions for NanoMips.

milica-lazarevic committed:
Select legal operand for "m" and "ZC" constraints

For "m" and "ZC" constraints, 9-bit offsets can be handled. Also, the
two least-significant bits should be cleared as required by ll and sc
instructions for NanoMips.

cme commented on llvm/lib/Target/Mips/MipsSEISelDAGToDAG.cpp:
```diff
@@ -1465,6 +1465,17 @@ SelectInlineAsmMemoryOperand(const SDValue &Op, unsigned ConstraintID,
+      }
+    }
+    [[fallthrough]];
   case InlineAsm::Constraint_o:
```
Wouldn't this apply to 'o' as a constraint too?

milica-lazarevic replied:
I am not completely sure if the "m" and "o" constraints should follow the same logic in the case of NanoMips architecture. It seems that GCC doesn't allow any offsets regarding the "m" constraint, while it allows 13-bit offsets for "o". If we want to adhere to this logic, we may need to separate conditions for Constraint_o and Constraint_m.

draganmladjenovic commented on llvm/lib/Target/Mips/MipsSEISelDAGToDAG.cpp:
```diff
@@ -1465,6 +1465,17 @@ SelectInlineAsmMemoryOperand(const SDValue &Op, unsigned ConstraintID,
+      // On NanoMips, they can only handle 9-bit offsets. Also, 2
+      // least-significant bits need to be cleared as required by LL and SC
+      // instructions.
+      if (selectAddrRegImm9(Op, Base, Offset)) {
```
This will verify that the offset is 4-byte aligned? Better move this to Constraint_o. Also on gcc m constraint doesn't allow offsets. While i disagree with that decision it is better to follow gcc here.

draganmladjenovic replied:
I see LL_NM and SC_NM are wrongly defined to use mem_simm9. Lets not make the same mistake here.

milica-lazarevic replied:
I agree the current solution isn't enough to ensure a 4-byte alignment. I will make sure to fix this.

Regarding the "o" constraint, it seems that to follow GCC logic, we may need to use something like this:
```
if (selectAddrFrameIndexOffset(Op, Base, Offset, 13)) {
        OutOps.push_back(Base);
        OutOps.push_back(Offset);
        return false;
}
```
Also, with larger offsets, GCC separates the constant in the following manner.
For example, if we have a constant 4098, GCC generates the code as follows:
```
        addiu	$a0,$a0,4096
	ll      $zero, 2($a0) 
```

Should we follow this convention, or 
```
        addiu	$a0,$a0,4098
	ll      $zero, 0($a0) 
```

would be sufficient?

draganmladjenovic replied:
For large offset just keep current clang behavior. For "o" I would thought that gcc would choose 9-bit signed offsets as that would be the lowers common denominator of "general" load/store instructions. @farazs-github Would this be an bug on gcc part? Or maybe non-symmetrical 12bit-unsigned 9bit negative range. Not sure.

draganmladjenovic commented on llvm/lib/Target/Mips/MipsSEISelDAGToDAG.cpp:
```diff
@@ -1505,6 +1516,15 @@ SelectInlineAsmMemoryOperand(const SDValue &Op, unsigned ConstraintID,
+      // On NanoMips, they can only handle 9-bit offsets. Also, 2
+      // least-significant bits need to be cleared as required by LL and SC
+      // instructions.
+      if (selectAddrRegImm9(Op, Base, Offset)) {
```
selectAddrFrameIndexOffset(Op, Base, Offset, 7, 2)?

milica-lazarevic committed:
Select operands for 'm', 'o' and 'ZC' constraints

In NanoMips, for the following constraints, these conditions need to be
fulfilled:
 'm'  - don't allow any offsets,
 'o'  - allow 13bit offsets,
 'ZC' - allow 9bit 4byte aligned offsets.

draganmladjenovic reviewed (changes requested):
LGTM. We had an request to include tests here as well. @farazs-github Does it make sense to follow gcc here for "o" constraint?

draganmladjenovic reviewed (approved):
LGTM
